apollo/lib/websocket_manager.py

```diff
@@ -0,0 +1,77 @@
+import uuid
+
+from fastapi import WebSocket
+from starlette.websockets import WebSocketDisconnect
+from typing import Dict
+
+from apollo.lib.singleton import Singleton
+
+
+class WebSocketManager(metaclass=Singleton):
+    def __init__(self):
+        self.connections: Dict[uuid.UUID, WebSocket] = {}
+
+    @staticmethod
+    async def _keep_connection_open(websocket: WebSocket):
+        try:
+            while True:
+                await websocket.receive_json()
+        except WebSocketDisconnect:
+            return
+
+    @staticmethod
+    async def _send_message(target_websocket, message):
+        await target_websocket.send_json(message)
+
+    @staticmethod
+    async def _wait_for_response(target_websocket: WebSocket,
+                                 message_id: uuid.UUID):
+        try:
+            while True:
+                response = await target_websocket.receive_json()
+                try:
+                    if response['message_id'] == str(message_id):
+                        return response
+                except KeyError:
+                    continue
+        except (WebSocketDisconnect, RuntimeError):
+            return
+
+    async def add_and_connect_websocket(self, client_id: uuid.UUID,
+                                        websocket: WebSocket):
+        self.connections[client_id] = websocket
+        await websocket.accept()
+        await websocket.send_json("Connection accepted")
+        await self._keep_connection_open(websocket)
+
+    async def send_message_and_wait_for_response(
+        self, target_websocket_id: uuid.UUID,
+        message: str
+    ):
+
+        message_id = uuid.uuid4()
+        target_websocket = self.connections[target_websocket_id]
+        await self._send_message(
+            target_websocket=target_websocket,
+            message={
+                'message_id': str(message_id),
+                'message': message
+            })
+        return await self._wait_for_response(
+            target_websocket=target_websocket,
+            message_id=message_id
+        )
+
+    async def close_and_remove_connection(self, websocket_id):
+        websocket = self.connections.pop(websocket_id)
+        try:
+            await websocket.send_json("Closing connection")
+            await websocket.close()
+        except RuntimeError:
+            # Connection already closed
```

Comment:
`RuntimeError` is pretty generic. Is there a way we can be more concrete?

Reply:
I dont think so?  https://github.com/encode/starlette/blob/4e2f301426d47cbe2bf3c5854e1a2276b9635a9e/starlette/websockets.py#L47

Reply:
Match (optionally partially) on the message.

Reply:
how ?

Reply:
As in an if statement and else throw the exception?

Reply:
```
except RuntimeError as e:
    if "disconnect message has been received" in str(e):
        return

    raise e
```

I noticed another `RuntimeError` catch earlier in this file  btw, which you might also want treat cautiously.